### ISSUE TITLE
Different SSL behaviour inside docker container

### ISSUE THREAD

**Issue by vidhill** [emqtt/emqttd#977]
#### Environment

- OS: Alpine
- Erlang/OTP: 
- EMQ: v2.0.7 

#### Description

Outside of  a docker container (on my local Ubuntu) I was able to configure ssl and verfiy that the certs worked

Inside the docker container I am seeing different behaviour for ssl, 
I am seeing this using the official dockerfile as well as my own..

I do have the relevant ports open

I tried building a docker image with ubuntu as the base image and then my SSL config worked, It would require me to do refactoring to use Ubuntu as a base image and it would be very heavy..

Is there some dependencies I am missing?

I tried installing `openssl` `ca-certificates` on my docker container but that didn't help

**Comment by vidhill** [comment]
Connecting via mqtts instead of ws I get the error...

```
17:08:44.255 [error] Client(172.17.0.1:43978): Parser failed for {function_clause,[{emqttd_parser,parse_utf,[<<1,0,239>>],[{file,"src/emqttd_parser.erl"},{line,206}]},{emqttd_parser,parse_frame,3,[{file,"src/emqttd_parser.erl"},{line,81}]},{emqttd_client,received,2,[{file,"src/emqttd_client.erl"},{line,259}]},{gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,615}]},{gen_server,handle_msg,5,[{file,"gen_server.erl"},{line,681}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,240}]}]}
17:08:44.255 [error] Client(172.17.0.1:43978): Error data: <<22,3,1,0,239,1,0,0,235,3,3,211,81,23,101,119,122,35,190,96,229,20,50,242,149,123,190,175,179,1,126,12,85,238,186,68,43,21,63,135,16,252,185,0,0,110,192,47,192,43,192,48,192,44,0,158,192,39,0,103,192,40,0,107,192,36,192,20,192,10,0,165,0,163,0,161,0,159,0,106,0,105,0,104,0,57,0,56,0,55,0,54,192,50,192,46,192,42,192,38,192,15,192,5,0,157,0,61,0,53,192,35,192,19,192,9,0,164,0,162,0,160,0,64,0,63,0,62,0,51,0,50,0,49,0,48,192,49,192,45,192,41,192,37,192,14,192,4,0,156,0,60,0,47,0,255,1,0,0,84,0,11,0,4,3,0,1,2,0,10,0,28,0,26,0,23,0,25,0,28,0,27,0,24,0,26,0,22,0,14,0,13,0,11,0,12,0,9,0,10,0,35,0,0,0,13,0,32,0,30,6,1,6,2,6,3,5,1,5,2,5,3,4,1,4,2,4,3,3,1,3,2,3,3,2,1,2,2,2,3,51,116,0,0>>
```

**Comment by emqplus** [comment]
@vidhill The SSL port is 8883.

**Comment by vidhill** [comment]
Yes, I know
this is happening  when I hit port 8883

When I Connect to the non secure connection:
`mqtt://mydomain` (not my actual domain)
Port `1883`
this works as expected

But connecting to:
`mqtts://mydomain` (not my actual domain)
Port `8883`

I get the error reported above

**Comment by vidhill** [comment]
Running EMQ locally on my host machine:
```
$ openssl s_client -connect localhost:8883
```
**Returns**
```
CONNECTED(00000003)
depth=0 CN = denghaiguideMacBook-Air.local, O = server
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 CN = denghaiguideMacBook-Air.local, O = server
verify error:num=21:unable to verify the first certificate
verify return:1
---
Certificate chain
 0 s:/CN=denghaiguideMacBook-Air.local/O=server
   i:/CN=MyTestCA
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIC9jCCAd6gAwIBAgIBATANBgkqhkiG9w0BAQsFADATMREwDwYDVQQDDAhNeVRl
c3RDQTAeFw0xNjEwMzEwNzE1NDVaFw0xNzEwMzEwNzE1NDVaMDkxJjAkBgNVBAMT
HWRlbmdoYWlndWlkZU1hY0Jvb2stQWlyLmxvY2FsMQ8wDQYDVQQKEwZzZXJ2ZXIw
ggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQC4Ena4vgWrzwUB0hGW1v0v
K986FhU5ZdYz5H5MGonfWwv89nR2DlftSDXEvKFyc2MT81GGm16VJv3mVpQJLuKA
xLBLY7a1zSrJdugXWy+mgJJTPW6KjTY4jPtfCl6x/yVr8YclVa8XO0JFzOme2LMV
Ylc/ixVEa66UpxRNrg5yWHS26KcB1lE3GLERoRBKF7nsyGqGY4X9TypBwglCVoqK
3dKVGwCvFur+oPnt/C5pwR6UmUV/Ppf1EaRD7Po+xcyJSeCvszG3FH4iHsDHnjLe
DR6lxouvMCb+aKJi9d0xowOjhbKoFMF179t4SVnptQeq+U6ui3cPKUjia7Zh1tZT
AgMBAAGjLzAtMAkGA1UdEwQCMAAwCwYDVR0PBAQDAgUgMBMGA1UdJQQMMAoGCCsG
AQUFBwMBMA0GCSqGSIb3DQEBCwUAA4IBAQB2jlDPiZfP/whsvvFn43g37QMwX5ST
Z5OpmEFnFjAH3ec0PPqPrKYEu00q5wEC+8L6uVH8FHOFf11JLH4wl11/C/mvE92D
qZtGG8KCnG2+rk5OJPGX+28Z+OnCZlXOjQ8qd2x5KtIW50JuXJ3cbDRHtF/TVanm
Exu+TCBeToNwbcU2sfQnbljkUTj4idUFz0pq3uvw3dA4R1J2foungPAYXSWcVhtb
RYtG8epIvkAyyUE5nY3kC05AUml6gSZkrJiYM5I1IJTX1lQ7Pv2yxRBZUtTx33rP
ccnsW6tbHTDBG8UDHx4LKHErdWFgCJWI81EUEcTip9g2zCOGTWKnpz+z
-----END CERTIFICATE-----
subject=/CN=denghaiguideMacBook-Air.local/O=server
issuer=/CN=MyTestCA
---
No client certificate CA names sent
---
SSL handshake has read 929 bytes and written 623 bytes
---
New, TLSv1/SSLv3, Cipher is AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : AES256-GCM-SHA384
    Session-ID: 343DF965E61FDF98FC56B7F95810B47D207340EEAD89653AC4C1324693363F0A
    Session-ID-ctx: 
    Master-Key: 3CA457415F7E2C3192DB10E382468E585FE63D59FC1E18EA01748680668AAA1618EB7799D770C60EB7E1641DBBC5DF67
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1491403534
    Timeout   : 300 (sec)
    Verify return code: 21 (unable to verify the first certificate)
---
closed
```

Running the official docker container (built locally)
```
docker run -p 18083:18083 -p 8883:8883 -p 1883:1883 --add-host my-test:127.0.0.1  emqttd:2.0.7 
```

Check ip set up correctly
```
$ ping my-test
PING my-test (172.17.0.2) 56(84) bytes of data.
^C
--- my-test ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 2014ms

```

**Running**
```
 openssl s_client -connect my-test:8883
```

**Returns**
```
CONNECTED(00000003)
140620099196568:error:140790E5:SSL routines:ssl23_write:ssl handshake failure:s23_lib.c:177:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 305 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID: 
    Session-ID-ctx: 
    Master-Key: 
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1491404134
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
---
```

**Comment by vidhill** [comment]
I think for some reason, EMQ is falling back to serving insecurely, perhaps

According to the [erlang ssl docs](http://erlang.org/doc/man/ssl.html)

> it falls back to insecure renegotiation if the peer does not support RFC 5746

?

**Comment by phanimahesh** [comment]
How are you setting config inside docker? `ADD`/`COPY` ed the config file?

What does `emqttd_ctl listeners` show inside docker?

**Comment by vidhill** [comment]
For above I'm using the https://github.com/emqtt/emq-docker (on the v2.0.7 tag)

In that the config file gets cloned down and copied as part of the Makefile for emqttd-relx

-I'll be able to give more details when I get into work in about an hour

**Comment by vidhill** [comment]
`emqttd_ctl listeners` outputs:

```
listener on mqtt:wss:8084
  acceptors       : 32
  max_clients     : 1000000
  current_clients : 0
  shutdown_count  : []
listener on mqtt:ws:8083
  acceptors       : 64
  max_clients     : 1000000
  current_clients : 0
  shutdown_count  : []
listener on mqtt:ssl:8883
  acceptors       : 32
  max_clients     : 500000
  current_clients : 0
  shutdown_count  : []
listener on mqtt:tcp:1883
  acceptors       : 64
  max_clients     : 1000000
  current_clients : 0
  shutdown_count  : []
listener on dashboard:http:18083
  acceptors       : 2
  max_clients     : 512
  current_clients : 4
  shutdown_count  : []

```

these are the permissions on the config files
```
-rw-r--r--    1 root     root           817 Apr  4 14:41 acl.conf
drwxr-xr-x    2 root     root          4096 Apr  4 15:46 certs
-rw-r--r--    1 root     root          7041 Apr  6 09:08 emq.conf
drwxr-xr-x    2 root     root          4096 Apr  4 14:41 plugins

```

And the certs
```
-rw-r--r--    1 root     root            38 Apr  4 14:37 README
-rw-r--r--    1 root     root          1021 Apr  4 14:37 cacert.pem
-rw-r--r--    1 root     root          1086 Apr  4 14:37 cert.pem
-rw-r--r--    1 root     root          1086 Apr  4 14:37 client-cert.pem
-rw-r--r--    1 root     root          1679 Apr  4 14:37 client-key.pem
-rw-r--r--    1 root     root          1679 Apr  4 14:37 key.pem

```

**Comment by vidhill** [comment]
Ok, I built `emq-docker` on master with is `2.1.0-rc.2` and when I run

`openssl s_client -connect my-test:8883`

I am successfully getting a secure connection.
-so it appears that the problem is specific to `v2.0.7`

I tried re-building `v2.0.7` with the newer version of Alpine ( simply changing   )
`ENV EMQ_VERSION=v2.1.0-rc.2` to `ENV EMQ_VERSION=v2.0.7p1`

And I see the same problem again..

**Comment by vidhill** [comment]
Issue was resolved by migrating to EMQ `v2.1.0`
